php/class-customize-snapshot-manager.php

```diff
@@ -381,6 +389,235 @@ function suspend_kses_for_snapshot_revision_restore() {
 	}
 
 	/**
+	 * Filter for displaying the Snapshot menu location values.
+	 *
+	 * @param array $menu_locations Default menu locations.
+	 * @return array Modified menu locations.
+	 */
+	public function filter_theme_mod_nav_menu_locations( $menu_locations ) {
+		if ( false === $this->snapshot->is_preview() ) {
+			return $menu_locations;
+		}
+
+		$values = $this->snapshot->values();
+		$locations = get_registered_nav_menus();
+
+		foreach ( $locations as $location => $name ) {
+			if ( isset( $values[ 'nav_menu_locations[' . $location . ']' ] ) ) {
+				$menu_locations[ $location ] = $values[ 'nav_menu_locations[' . $location . ']' ];
+			}
+		}
+
+		return $menu_locations;
+	}
+
+	/**
+	 * Filter wp_get_nav_menus() to load Snapshot values.
+	 *
+	 * @see wp_get_nav_menus()
+	 *
+	 * @param array $menus Array of menus.
+	 * @return array Modified array of menus.
+	 */
+	public function filter_wp_get_nav_menus( $menus ) {
+		if ( false === $this->snapshot->is_preview() ) {
+			return $menus;
+		}
+
+		$values = $this->snapshot->values();
+		$removed = array();
+
+		foreach ( $values as $setting_id => $value ) {
+			if ( preg_match( '/^nav_menu\[(?P<id>-?\d+)\]$/', $setting_id, $matches ) ) {
+				if ( false !== $value ) {
+					$menus[] = $this->get_nav_menu_object( $matches['id'], $value );
+				} else {
+					$removed[] = intval( $matches['id'] );
+				}
+			}
+		}
+
+		if ( ! empty( $removed ) && ! empty( $menus ) ) {
+			foreach ( $menus as $key => $term ) {
+				if ( in_array( $term->term_id, $removed, true ) ) {
+					unset( $menus[ $key ] );
+				}
+			}
+		}
+
+		return array_values( $menus );
+	}
+
+	/**
+	 * Filter wp_get_nav_menu_items() to load Snapshot values.
+	 *
+	 * @see wp_get_nav_menu_items()
+	 *
+	 * @param array  $items An array of menu item post objects.
+	 * @param object $menu  The menu object.
+	 * @param array  $args  An array of arguments used to retrieve menu item objects.
+	 * @return array Array of menu items.
+	 */
+	function filter_wp_get_nav_menu_items( $items, $menu, $args ) {
+		if ( false === $this->snapshot->is_preview() ) {
+			return $items;
+		}
+
+		$values = $this->snapshot->values();
+		$removed = array();
+
+		foreach ( $values as $setting_id => $value ) {
+			if ( preg_match( '/^nav_menu_item\[(?P<id>-?\d+)\]$/', $setting_id, $matches ) && false === $value ) {
+				$removed[] = intval( $matches['id'] );
+			}
+		}
+
+		if ( ! empty( $removed ) && ! empty( $items ) ) {
+			foreach ( $items as $key => $post ) {
+				if ( in_array( $post->ID, $removed, true ) ) {
+					unset( $items[ $key ] );
+				}
+			}
+		}
+
+		foreach ( $values as $setting_id => $item ) {
+			if ( preg_match( \WP_Customize_Nav_Menu_Item_Setting::ID_PATTERN, $setting_id, $matches ) ) {
+				if ( (int) $menu->term_id === (int) $item['nav_menu_term_id'] ) {
+					$item['post_id'] = intval( $matches['id'] );
+					$items[] = $this->value_as_wp_post_nav_menu_item( (object) $item );
```

Comment:
I believe we can re-use `WP_Customize_Nav_Menu_Item_Setting::value_as_wp_post_nav_menu_item()` here.

```php
if ( preg_match( \WP_Customize_Nav_Menu_Item_Setting::ID_PATTERN, $setting_id, $matches ) ) {
    $wp_customize->register_dynamic_settings( array( $setting_id ) );
    $nav_menu_item_setting = $wp_customize->get_setting( $setting_id );
    if ( $nav_menu_item_setting && (int) $menu->term_id === (int) $item['nav_menu_term_id'] ) {
        $item['post_id'] = intval( $matches['id'] );
        $items[] = $nav_menu_item_setting->value_as_wp_post_nav_menu_item(); // <========
    }
}
```